src/ComputeSharp.D2D1/Shaders/Interop/Effects/PixelShaderEffect.ID2D1DrawTransformMethods.cs

```diff
@@ -211,16 +225,32 @@ public static int MapInputRectsToOutputRect(PixelShaderEffect* @this, RECT* inpu
                 // transform. In this case, the target area will be the output node anyway.
                 outputRect->MakeD2D1Infinite();
 
-                @this->inputRect = default;
-
                 *outputOpaqueSubRect = default;
             }
             else
             {
-                // Default mapping
-                *outputRect = inputRects[0];
+                // If at least one input is present and no custom mapper is available, apply the default
+                // mapping. In this case, the output rect should be the union of the input rects for all
+                // the input shader inputs, or infinite if there are no simple inputs. The input rects
+                // for complex inputs, if present, are not needed in this scenario, so they're ignored.
+                RECT? unionOfSimpleRects = null;
 
-                @this->inputRect = inputRects[0];
+                for (uint i = 0; i < inputRectCount; i++)
+                {
+                    if (@this->inputTypes[i] == D2D1PixelShaderInputType.Simple)
+                    {
+                        unionOfSimpleRects = (unionOfSimpleRects ?? default).Union(inputRects[i]);
```

Comment:
using `default` is not actually correct here, because then you're taking the union of `(0,0,0,0)` and another rect.

The union of `(0,0,0,0)` and `(x=5,y=5,w=1,h=1)` will be `(x=0,y=0,w=6,h=6)`, for instance. In my code I used `Empty` which is a special rectangle that equals `(x = int.max, y=int.max, w=int.min, h=int.min)` which behaves correctly/specially for things like union and intersect (IIUC).

Really you want `unionOfSimpleRects = unionOfSimpleRects.HasValue ? unionOfSimpleRects.Union(inputRects[i]) : inputRects[i];` (might be a better way to express that but you get the idea).

Alternately, have a `Union(RECT? rect1, RECT rect2)` overload that handles this

Reply:
You could also define a `RECT.Empty = { int.MaxValue, int.MaxValue, int.MinValue, int.MinValue }` and use that as your "null" value. Double check that it works when `Union()`-ing with a regular RECT though

Reply:
I'm not sure I'm following, I'm working on `RECT` here with is LTRB, not XYWH, so wouldn't the union still work correctly in all cases? As in, it's just taking the minimum of LT, and the maximum of RB, so thee should be no concerns about the height and the width being tracked, no? 🤔

Anyway I can still rewrite that for clarity, sure 🙂

Reply:
Fixed in https://github.com/Sergio0694/ComputeSharp/pull/265/commits/f2be773ab6055fef8380ba04de3a9a113f8391e8.